src/libraries/System.Text.Json/src/System/Text/Json/Serialization/IgnoreReferenceResolver.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class IgnoreReferenceResolver : ReferenceResolver
+    {
+        // The stack of references on the branch of the current object graph, used to detect reference cycles.
+        private Stack<ReferenceEqualsWrapper>? _stackForCycleDetection;
+
+        internal override void PopReferenceForCycleDetection()
+        {
+            if (_stackForCycleDetection != null)
+            {
+                _stackForCycleDetection.Pop();
+            }
+        }
+
+        internal override bool ContainsReferenceForCycleDetection(object value)
+            => _stackForCycleDetection?.Contains(new ReferenceEqualsWrapper(value)) ?? false;
+
+        internal override void PushReferenceForCycleDetection(object value)
+        {
+            var wrappedValue = new ReferenceEqualsWrapper(value);
```

Comment:
`Stack<T>.Contains` uses an `object.Equals(object)` comparison and there is no way to tell the method to use a different comparison (in our case, we need reference equality comparison). 

Therefore we use `Stack<ReferenceEqualsWrapper>` where ReferenceEqualsWrapper is a thin struct wrapper for the objects in the graph that overrides `object.Equals` to call `object.ReferenceEquals` using the underlying `_object` in the wrapper.

Reply:
Opposed to this, we can avoid forcing reference equality and rely on object.Equals, which is what Newtonsoft does (see https://github.com/JamesNK/Newtonsoft.Json/issues/401#issuecomment-77711185).

Also, Newtonsoft offers [`JsonSerializerSettings.EqualityComparer`](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializerSettings_EqualityComparer.htm) which can be used to allow users to specify their own equality comparison for this case.

Reply:
I prefer the current choice of forcing reference equality for this feature since it aligns with Preserve, which uses `ReferenceEqualityComparer`. 

If requested by users, we can consider adding an API to a future extensibility model which allows users to specify an equality comparer to use with IgnoreCycles.

Reply:
Got it, thanks. This doesn't affect behavior when this feature is off (happy path), but are there any perf issues with struct copies when inserting/checking/extracting from the stack?

Reply:
We only pass reference-types to the stack, in the case of boxed structs, there is no copy semantics applying given that we never unbox.